### PR TITLE
Add username to menu

### PR DESCRIPTION
**Pull request by miyamoto-san**

### CODE REVIEW

**Review by kaj — CHANGES_REQUESTED** [review]
Please either introduce a format checker / linter in a separate PR or just revert the formatting changes in this PR.  This way it is impossible to see the actual code changes that adds support for login.

File: app.js [hunk 1]
```diff
@@ -1,103 +1,147 @@
...
+  }
+
+  serviceValidateUrl.searchParams.set("ticket", req.query.ticket);
+  serviceValidateUrl.searchParams.set("service", serviceUrl);
+
+  const { body } = await got(serviceValidateUrl);
+
+  if (validate(body) === true) {
+    userName = parse(body)['cas:serviceResponse']['cas:authenticationSuccess']['cas:user'];
+  }
```
Comment (kaj) [inline app.js:112]: Why validate before parsing?  We still needs to handle an exception if we get a valid xml document without the expected fields.

File: app.js [hunk 2]
```diff
@@ -1,103 +1,147 @@
...
+  res.setHeader("Expires", addDays(new Date(), 180).toUTCString());
+  res.send(menuCssData);
+});
+
+app.get("/kpm/kpm.js", async (req, res) => {
+  const baseUrl = `${process.env.SERVER_HOST_URL}/kpm/`;
+  const cssUrl = `${baseUrl}${menuCssName}`;
+  let userName;
+
+  const serviceUrl = "http://localdev.kth.se/kpm/kpm.js";
```
Comment (kaj) [inline app.js:91]: This should be `${aseUrl}/kpm.js`.

File: app.js [hunk 3]
```diff
@@ -1,103 +1,147 @@
...
+  const loginUrl = new URL("https://login-r.referens.sys.kth.se/login");
+  const serviceValidateUrl = new URL(
+    "https://login-r.referens.sys.kth.se/serviceValidate"
+  );
```
Comment (kaj) [inline app.js:95]: The login houst needs to come from the environment.

File: app.js [hunk 4]
```diff
@@ -1,103 +1,147 @@
...
-app.listen(3000, () => {
-  log.info('Starting app KPM in port 3000')
-})
+http.createServer(app).listen(80);
+https
+  .createServer(
+    {
+      key: readFileSync("./certs/key.pem"),
+      cert: readFileSync("./certs/cert.pem"),
+      passphrase: "lolo",
```
Comment (kaj) [inline app.js:143]: We can't put the cert in the docker image, so I guess we need to get it through then environment and not from files.

**Review by kaj — COMMENTED** [review]
Looks like a reasonable first step.  I have some issues which needs to be adressed to get it running in stage.

I also think the actual login should not be handled in the request for `kpm.js` (which I think should be cacheable) but in a separate request for the content of the menu bar.  But having it _somewhere_ is a reasonable first step.

File: app.js (hunk 3 above)
Comment (kaj) [reply app.js:95]: I guess / hope it is enough to have one environment variable with the host name and build both `loginUrl` and `serviceValidateUrl` from that.  :-)

File: app.js (hunk 1 above)
Comment (miyamoto-san) [reply app.js:112]: Yes, this is what I wanted to discuss!

File: app.js (hunk 4 above)
Comment (exacs) [reply app.js:143]: These files are only needed when running the app locally. Remember that in stage/prod, app is behind a reverse proxy that handles the HTTPS stuff.

If that is the case, I'd say we might need two "start" files or something

**Comment by kaj** [comment]
Requests that has a `ticket` query parameter should always get a redirect.  But that requires some kind of session handling, so maybe it's for the next PR.

File: src/server.js [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+  return res.body;
+}
+
+app.use("/kpm/dist", express.static("dist"));
+
+app.get("/kpm/_monitor", (req, res) => {
+  res.setHeader("Content-Type", "text/plain");
+  let version = "unknown";
+  try {
+    const info = require("./buildinfo.js");
```
Comment (kaj) [inline src/server.js:70]: I think this needs to be `"../buildinfo.js"`, now that this function is moved into the `src` directory.

**Review by kaj — CHANGES_REQUESTED** [review]

**Pull request merged**